src/Interpreters/Streaming/Aggregator.cpp

```diff
@@ -1274,283 +1366,70 @@ Block Aggregator::convertOneBucketToBlockImpl(
-
-    APPLY_FOR_VARIANTS_TIME_BUCKET_TWO_LEVEL(M)
-#undef M
-    else
-        throw Exception("Unknown aggregated data variant.", ErrorCodes::UNKNOWN_AGGREGATED_DATA_VARIANT);
+    size_t max_temporary_block_size_rows = 0;
+    size_t max_temporary_block_size_bytes = 0;
 
-    return block;
-}
+    auto update_max_sizes = [&](const Block & block)
+    {
+        size_t block_size_rows = block.rows();
+        size_t block_size_bytes = block.bytes();
 
-Block Aggregator::mergeAndConvertOneBucketToBlock(
-    ManyAggregatedDataVariants & variants, bool final, ConvertAction action, size_t bucket) const
-{
-    auto prepared_data_ptr = prepareVariantsToMerge(variants);
-    if (prepared_data_ptr->empty())
-        return {};
+        if (block_size_rows > max_temporary_block_size_rows)
+            max_temporary_block_size_rows = block_size_rows;
+        if (block_size_bytes > max_temporary_block_size_bytes)
+            max_temporary_block_size_bytes = block_size_bytes;
+    };
 
-    auto & merged_data = *prepared_data_ptr->at(0);
-    auto method = merged_data.type;
-    Arena * arena = merged_data.aggregates_pool;
-    bool clear_states = shouldClearStates(action, final);
-    Block block;
+    for (auto bucket : method.data.buckets())
+    {
+        Block block = convertOneBucketToBlockImpl(data_variants, method, data_variants.aggregates_pool, false, false, bucket);
+        out.write(block);
+        update_max_sizes(block);
+    }
 
-    if (false) {} // NOLINT
-#define M(NAME) \
-    else if (method == AggregatedDataVariants::Type::NAME) \
-    { \
-        mergeBucketImpl<decltype(merged_data.NAME)::element_type>(*prepared_data_ptr, final, clear_states, bucket, arena); \
-        block = convertOneBucketToBlockImpl(merged_data, *merged_data.NAME, arena, final, clear_states, bucket); \
+    if (params.overflow_row)
+    {
+        Block block = prepareBlockAndFillWithoutKey(data_variants, false, true, false);
+        out.write(block);
+        update_max_sizes(block);
     }
 
-    APPLY_FOR_VARIANTS_ALL_TWO_LEVEL(M)
-#undef M
+    /// Pass ownership of the aggregate functions states:
+    /// `data_variants` will not destroy them in the destructor, they are now owned by ColumnAggregateFunction objects.
+    data_variants.aggregator = nullptr;
 
-    return block;
+    LOG_DEBUG(log, "Max size of temporary block: {} rows, {}.", max_temporary_block_size_rows, ReadableSize(max_temporary_block_size_bytes));
 }
 
-BlocksList
-Aggregator::mergeAndConvertToBlocks(ManyAggregatedDataVariants & data_variants, bool final, ConvertAction action, size_t max_threads) const
-{
-    auto prepared_data_ptr = prepareVariantsToMerge(data_variants);
-    if (prepared_data_ptr->empty())
-        return {};
 
-    if (unlikely(params.overflow_row))
-        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Overflow row processing is not implemented in streaming aggregation");
+bool Aggregator::checkLimits(size_t result_size, bool & no_more_keys) const
```

Comment:
For now, probably we can disable limit checking since for streaming processing, it is hard to limit a size since it is long running

Reply:
can do it in separate PR